File: electron/providers/copilot-auth.ts
/**
 * GitHub Copilot OAuth Authentication
 * Implements Device Flow for GitHub OAuth
 */

import { logger } from "../utils/logger";
import { storage } from "../services/storage";

// ── Constants ─────────────────────────────────────────────

// GitHub Copilot client ID (public, from VS Code Copilot extension)
const COPILOT_CLIENT_ID = "Iv1.b507a08c87ecfe98";

const GITHUB_DEVICE_CODE_URL = "https://github.com/login/device/code";
const GITHUB_TOKEN_URL = "https://github.com/login/oauth/access_token";
const GITHUB_USER_URL = "https://api.github.com/user";
const COPILOT_TOKEN_URL = "https://api.github.com/copilot_internal/v2/token";

// ── Types ─────────────────────────────────────────────────

export interface DeviceCodeResponse {
  device_code: string;
  user_code: string;
  verification_uri: string;
  expires_in: number;
  interval: number;
}

export interface TokenResponse {
  access_token: string;
  token_type: string;
  scope: string;
  error?: string;
  error_description?: string;
}

export interface CopilotToken {
  token: string;
  expires_at: number;
}

export interface GitHubUser {
  login: string;
  name: string | null;
  avatar_url: string;
}

// ── Helper Functions ──────────────────────────────────────

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ── Device Flow Implementation ────────────────────────────

/**
 * Step 1: Request device code from GitHub
 */
export async function initiateDeviceFlow(): Promise<DeviceCodeResponse> {
  logger.info("copilot-auth", "Initiating device flow");

  const response = await fetch(GITHUB_DEVICE_CODE_URL, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify({
      client_id: COPILOT_CLIENT_ID,
      scope: "read:user",
    }),
  });

  if (!response.ok) {
    throw new Error(`Device code request failed: ${response.status}`);
  }

  const data = (await response.json()) as DeviceCodeResponse;
  logger.info("copilot-auth", `Device code received, user code: ${data.user_code}`);

  return data;
}

/**
 * Step 2: Poll for token after user authorizes
 */
export async function pollForToken(
  deviceCode: string,
  interval: number,
  expiresIn: number
): Promise<string> {
  const startTime = Date.now();
  const expiryTime = startTime + expiresIn * 1000;

  logger.info("copilot-auth", "Polling for token...");

  while (Date.now() < expiryTime) {
    await sleep(interval * 1000);

    const response = await fetch(GITHUB_TOKEN_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify({
        client_id: COPILOT_CLIENT_ID,
        device_code: deviceCode,
        grant_type: "urn:ietf:params:oauth:grant-type:device_code",
      }),
    });

    const data = (await response.json()) as TokenResponse;

    if (data.access_token) {
      logger.info("copilot-auth", "Token received successfully");
      return data.access_token;
    }

    if (data.error === "authorization_pending") {
      // User hasn't authorized yet, keep polling
      continue;
    }

    if (data.error === "slow_down") {
      // Increase polling interval
      interval += 5;
      continue;
    }

    if (data.error === "expired_token") {
      throw new Error("Device code expired. Please try again.");
    }

    if (data.error === "access_denied") {
      throw new Error("Authorization was denied by user.");
    }

    if (data.error) {
      throw new Error(data.error_description || data.error);
    }
  }

  throw new Error("Polling timed out. Please try again.");
}

/**
 * Get GitHub user info from access token
 */
export async function getGitHubUser(accessToken: string): Promise<GitHubUser> {
  const response = await fetch(GITHUB_USER_URL, {
    headers: {
      Authorization: `Bearer ${accessToken}`,
      Accept: "application/json",
    },
  });

  if (!response.ok) {
    throw new Error(`Failed to get user info: ${response.status}`);
  }

  return (await response.json()) as GitHubUser;
}

/**
 * Get Copilot API token from GitHub access token
 */
export async function getCopilotToken(accessToken: string): Promise<CopilotToken> {
  const response = await fetch(COPILOT_TOKEN_URL, {
    headers: {
      Authorization: `token ${accessToken}`,
      Accept: "application/json",
      "Editor-Version": "vscode/1.85.0",
      "Editor-Plugin-Version": "copilot/1.0.0",
    },
  });

  if (!response.ok) {
    const text = await response.text();
    logger.error("copilot-auth", `Copilot token request failed: ${response.status} - ${text}`);
    throw new Error(`Failed to get Copilot token: ${response.status}`);
  }

  return (await response.json()) as CopilotToken;
}

// ── Token Storage ─────────────────────────────────────────

/**
 * Save tokens to storage
 */
export function saveTokens(
  accessToken: string,
  username: string,
  expiresAt?: number
): void {
  const settings = storage.getAppSettings();

  // Note: In production, tokens should be encrypted
  // For now, we store them in settings (which uses SQLite)
  const updatedSettings: Partial<typeof settings> = {
    providers: {
      defaultProvider: settings.providers?.defaultProvider ?? "claude",
      claude: settings.providers?.claude ?? { enabled: true, defaultModel: "sonnet" },
      copilot: {
        enabled: true,
        authenticated: true,
        accessToken,
        tokenExpiry: expiresAt,
        username,
        defaultModel: settings.providers?.copilot?.defaultModel ?? "gpt-4.1",
      },
    },
  };

  storage.updateAppSettings(updatedSettings);
  logger.info("copilot-auth", `Tokens saved for user: ${username}`);
}

/**
 * Get stored access token
 */
export function getStoredToken(): string | null {
  const settings = storage.getAppSettings();
  return settings.providers?.copilot?.accessToken ?? null;
}

/**
 * Check if token is expired
 */
export function isTokenExpired(): boolean {
  const settings = storage.getAppSettings();
  const expiry = settings.providers?.copilot?.tokenExpiry;

  if (!expiry) return false;

  // Add 5 minute buffer
  return Date.now() > expiry - 5 * 60 * 1000;
}

/**
 * Get stored username
 */
export function getStoredUsername(): string | null {
  const settings = storage.getAppSettings();
  return settings.providers?.copilot?.username ?? null;
}

/**
 * Check authentication status
 */
export function isAuthenticated(): boolean {
  const settings = storage.getAppSettings();
  return settings.providers?.copilot?.authenticated === true &&
         !!settings.providers?.copilot?.accessToken;
}

/**
 * Clear all stored tokens (logout)
 */
export function clearTokens(): void {
  const settings = storage.getAppSettings();

  const updatedSettings: Partial<typeof settings> = {
    providers: {
      defaultProvider: settings.providers?.defaultProvider ?? "claude",
      claude: settings.providers?.claude ?? { enabled: true, defaultModel: "sonnet" },
      copilot: {
        enabled: false,
        authenticated: false,
        accessToken: undefined,
        refreshToken: undefined,
        tokenExpiry: undefined,
        username: undefined,
        defaultModel: settings.providers?.copilot?.defaultModel ?? "gpt-4.1",
      },
    },
  };

  storage.updateAppSettings(updatedSettings);
  logger.info("copilot-auth", "Tokens cleared");
}

// ── Public API ────────────────────────────────────────────

export const copilotAuth = {
  initiateDeviceFlow,
  pollForToken,
  getGitHubUser,
  getCopilotToken,
  saveTokens,
  getStoredToken,
  getStoredUsername,
  isAuthenticated,
  isTokenExpired,
  clearTokens,
};

File: src-tauri/src/providers/copilot_auth.rs
use anyhow::{bail, Result};
use rusqlite::Connection;
use serde::{Deserialize, Serialize};
use tracing::{info, error};

use crate::db::models::{ClaudeProviderSettings, CopilotProviderSettings, ProvidersSettings, ProviderType};
use crate::db::queries::settings;

// ── Constants ────────────────────────────────────────────────

/// GitHub Copilot client ID (public, from VS Code Copilot extension).
pub const COPILOT_CLIENT_ID: &str = "Iv1.b507a08c87ecfe98";

pub const GITHUB_DEVICE_CODE_URL: &str = "https://github.com/login/device/code";
pub const GITHUB_TOKEN_URL: &str = "https://github.com/login/oauth/access_token";
pub const GITHUB_USER_URL: &str = "https://api.github.com/user";
pub const COPILOT_TOKEN_URL: &str = "https://api.github.com/copilot_internal/v2/token";

// ── Types ────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeviceCodeResponse {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    pub expires_in: u64,
    pub interval: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenResponse {
    #[serde(default)]
    pub access_token: Option<String>,
    #[serde(default)]
    pub token_type: Option<String>,
    #[serde(default)]
    pub scope: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
    #[serde(default)]
    pub error_description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CopilotToken {
    pub token: String,
    pub expires_at: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GitHubUser {
    pub login: String,
    pub name: Option<String>,
    pub avatar_url: String,
}

// ── Device Flow Implementation ───────────────────────────────

/// Step 1: Request device code from GitHub.
pub async fn initiate_device_flow() -> Result<DeviceCodeResponse> {
    info!("copilot-auth: initiating device flow");

    let client = reqwest::Client::new();
    let response = client
        .post(GITHUB_DEVICE_CODE_URL)
        .header("Content-Type", "application/json")
        .header("Accept", "application/json")
        .json(&serde_json::json!({
            "client_id": COPILOT_CLIENT_ID,
            "scope": "read:user",
        }))
        .send()
        .await?;

    if !response.status().is_success() {
        bail!("Device code request failed: {}", response.status());
    }

    let data: DeviceCodeResponse = response.json().await?;
    info!("copilot-auth: device code received, user code: {}", data.user_code);

    Ok(data)
}

/// Step 2: Poll for token after user authorizes.
pub async fn poll_for_token(
    device_code: &str,
    mut interval: u64,
    expires_in: u64,
) -> Result<String> {
    let start = std::time::Instant::now();
    let expiry = std::time::Duration::from_secs(expires_in);

    info!("copilot-auth: polling for token...");

    let client = reqwest::Client::new();

    while start.elapsed() < expiry {
        tokio::time::sleep(std::time::Duration::from_secs(interval)).await;

        let response = client
            .post(GITHUB_TOKEN_URL)
            .header("Content-Type", "application/json")
            .header("Accept", "application/json")
            .json(&serde_json::json!({
                "client_id": COPILOT_CLIENT_ID,
                "device_code": device_code,
                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
            }))
            .send()
            .await?;

        let data: TokenResponse = response.json().await?;

        if let Some(access_token) = data.access_token {
            if !access_token.is_empty() {
                info!("copilot-auth: token received successfully");
                return Ok(access_token);
            }
        }

        match data.error.as_deref() {
            Some("authorization_pending") => continue,
            Some("slow_down") => {
                interval += 5;
                continue;
            }
            Some("expired_token") => bail!("Device code expired. Please try again."),
            Some("access_denied") => bail!("Authorization was denied by user."),
            Some(err) => {
                let desc = data.error_description.unwrap_or_default();
                bail!("{}", if desc.is_empty() { err.to_string() } else { desc });
            }
            None => continue,
        }
    }

    bail!("Polling timed out. Please try again.")
}

/// Get GitHub user info from access token.
pub async fn get_github_user(access_token: &str) -> Result<GitHubUser> {
    let client = reqwest::Client::new();
    let response = client
        .get(GITHUB_USER_URL)
        .header("Authorization", format!("Bearer {}", access_token))
        .header("Accept", "application/json")
        .send()
        .await?;

    if !response.status().is_success() {
        bail!("Failed to get user info: {}", response.status());
    }

    let user: GitHubUser = response.json().await?;
    Ok(user)
}

/// Get Copilot API token from GitHub access token.
pub async fn get_copilot_token(access_token: &str) -> Result<CopilotToken> {
    let client = reqwest::Client::new();
    let response = client
        .get(COPILOT_TOKEN_URL)
        .header("Authorization", format!("token {}", access_token))
        .header("Accept", "application/json")
        .header("Editor-Version", "vscode/1.85.0")
        .header("Editor-Plugin-Version", "copilot/1.0.0")
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response.text().await.unwrap_or_default();
        error!("copilot-auth: copilot token request failed: {} - {}", status, text);
        bail!("Failed to get Copilot token: {}", status);
    }

    let token: CopilotToken = response.json().await?;
    Ok(token)
}

// ── Token Storage ────────────────────────────────────────────

/// Save tokens to the app settings (SQLite).
pub fn save_tokens(conn: &Connection, access_token: &str, username: &str) {
    let current_settings = settings::get_app_settings(conn).unwrap_or_default();

    let providers = current_settings.providers.unwrap_or_else(|| ProvidersSettings {
        default_provider: ProviderType::Claude,
        claude: ClaudeProviderSettings {
            enabled: true,
            cli_path: None,
            default_model: "sonnet".to_string(),
        },
        copilot: CopilotProviderSettings {
            enabled: false,
            authenticated: false,
            access_token: None,
            refresh_token: None,
            token_expiry: None,
            default_model: "gpt-4.1".to_string(),
            username: None,
        },
    });

    let updated_providers = ProvidersSettings {
        default_provider: providers.default_provider,
        claude: providers.claude,
        copilot: CopilotProviderSettings {
            enabled: true,
            authenticated: true,
            access_token: Some(access_token.to_string()),
            refresh_token: None,
            token_expiry: None,
            default_model: providers.copilot.default_model,
            username: Some(username.to_string()),
        },
    };

    let update = serde_json::json!({
        "providers": serde_json::to_value(&updated_providers).unwrap_or_default(),
    });

    if let Err(e) = settings::update_app_settings(conn, &update) {
        error!("copilot-auth: failed to save tokens: {}", e);
    } else {
        info!("copilot-auth: tokens saved for user: {}", username);
    }
}

/// Get stored access token from app settings.
pub fn get_stored_token(conn: &Connection) -> Option<String> {
    let app_settings = settings::get_app_settings(conn).ok()?;
    app_settings.providers?.copilot.access_token
}

/// Check if the user is authenticated with Copilot.
pub fn is_authenticated(conn: &Connection) -> bool {
    let app_settings = match settings::get_app_settings(conn) {
        Ok(s) => s,
        Err(_) => return false,
    };
    match app_settings.providers {
        Some(p) => p.copilot.authenticated && p.copilot.access_token.is_some(),
        None => false,
    }
}

/// Clear all stored tokens (logout).
pub fn clear_tokens(conn: &Connection) {
    let current_settings = settings::get_app_settings(conn).unwrap_or_default();

    let providers = current_settings.providers.unwrap_or_else(|| ProvidersSettings {
        default_provider: ProviderType::Claude,
        claude: ClaudeProviderSettings {
            enabled: true,
            cli_path: None,
            default_model: "sonnet".to_string(),
        },
        copilot: CopilotProviderSettings {
            enabled: false,
            authenticated: false,
            access_token: None,
            refresh_token: None,
            token_expiry: None,
            default_model: "gpt-4.1".to_string(),
            username: None,
        },
    });

    let updated_providers = ProvidersSettings {
        default_provider: providers.default_provider,
        claude: providers.claude,
        copilot: CopilotProviderSettings {
            enabled: false,
            authenticated: false,
            access_token: None,
            refresh_token: None,
            token_expiry: None,
            default_model: providers.copilot.default_model,
            username: None,
        },
    };

    let update = serde_json::json!({
        "providers": serde_json::to_value(&updated_providers).unwrap_or_default(),
    });

    if let Err(e) = settings::update_app_settings(conn, &update) {
        error!("copilot-auth: failed to clear tokens: {}", e);
    } else {
        info!("copilot-auth: tokens cleared");
    }
}
